utils/zerovec/src/map/mod.rs

```diff
@@ -0,0 +1,273 @@
+
+mod kv;
+#[cfg(feature = "serde")]
+mod serde;
+mod vecs;
+
+pub use kv::ZeroMapKV;
+pub use vecs::ZeroVecLike;
+
+/// A zero-copy map datastructure, built on sorted binary-searchable [`ZeroVec`]
+/// and [`VarZeroVec`](crate::VarZeroVec).
+///
+/// This type, like [`ZeroVec`] and [`VarZeroVec`](crate::VarZeroVec), is able to zero-copy
+/// deserialize from appropriately formatted byte buffers. It is internally copy-on-write, so it can be mutated
+/// afterwards as necessary.
+pub struct ZeroMap<'a, K, V>
+where
+    K: ZeroMapKV<'a>,
+    V: ZeroMapKV<'a>,
+{
+    keys: K::Container,
+    values: V::Container,
+}
+
+impl<'a, K, V> Default for ZeroMap<'a, K, V>
+where
+    K: ZeroMapKV<'a>,
+    V: ZeroMapKV<'a>,
+{
+    fn default() -> Self {
+        Self {
+            keys: K::Container::new(),
+            values: V::Container::new(),
+        }
+    }
+}
+
+impl<'a, K, V> ZeroMap<'a, K, V>
+where
+    K: ZeroMapKV<'a>,
+    V: ZeroMapKV<'a>,
+{
+    /// Construct a new [`ZeroMap`]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Construct a new [`ZeroMap`] with a given capacity
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            keys: K::Container::with_capacity(capacity),
+            values: V::Container::with_capacity(capacity),
+        }
+    }
+
+    /// The number of elements in the [`ZeroMap`]
+    pub fn len(&self) -> usize {
+        self.values.len()
+    }
+
+    /// Whether the [`ZeroMap`] is empty
+    pub fn is_empty(&self) -> bool {
+        self.values.len() == 0
+    }
+
+    /// Remove all elements from the [`ZeroMap`]
+    pub fn clear(&mut self) {
+        self.keys.clear();
+        self.values.clear();
+    }
+
+    /// Reserve capacity for `additional` more elements to be inserted into
+    /// the [`ZeroMap`] to avoid frequent reallocations.
+    ///
+    /// See [`Vec::reserve()`] for more information.
+    pub fn reserve(&mut self, additional: usize) {
+        self.keys.reserve(additional);
+        self.values.reserve(additional);
+    }
+
+    /// Get the value associated with `key`, if it exists.
+    ///
+    /// ```rust
+    /// use zerovec::ZeroMap;
+    ///
+    /// let mut map = ZeroMap::new();
+    /// map.insert(1, "one".to_owned());
+    /// map.insert(2, "two".to_owned());
+    /// assert_eq!(map.get(&1), Some("one"));
+    /// assert_eq!(map.get(&3), None);
+    /// ```
+    pub fn get(&self, key: &K::NeedleType) -> Option<&V::GetType> {
+        let index = self.keys.binary_search(key).ok()?;
+        self.values.get(index)
+    }
+
+    /// Returns whether `key` is contained in this map
+    ///
+    /// ```rust
+    /// use zerovec::ZeroMap;
+    ///
+    /// let mut map = ZeroMap::new();
+    /// map.insert(1, "one".to_owned());
+    /// map.insert(2, "two".to_owned());
+    /// assert_eq!(map.contains_key(&1), true);
+    /// assert_eq!(map.contains_key(&3), false);
+    /// ```
+    pub fn contains_key(&self, key: &K) -> bool {
```

Comment:
Question/Suggestion: Why not take `&K::NeedleType` here directly?

Reply:
good catch